pkg/util/workqueue/delaying_queue.go

```diff
@@ -0,0 +1,172 @@
+	// is beyond the time we're already waiting for, we don't add a new timer thread
+	nextCheckTime time.Time
+	// nextCheckLock serializes access to the notification time
+	nextCheckLock sync.Mutex
+}
+
+// waitFor holds the data to add and the time it should be added
+type waitFor struct {
+	data    t
+	readyAt time.Time
+}
+
+// ShutDown gives a way to shut off this queue
+func (q *delayingType) ShutDown() {
+	q.Type.ShutDown()
+	q.waitingCond.Broadcast()
+}
+
+func (q *delayingType) AddAfter(item interface{}, duration time.Duration) {
+	q.waitingLock.Lock()
+	defer q.waitingLock.Unlock()
+	waitEntry := waitFor{data: item, readyAt: q.clock.Now().Add(duration)}
+
+	insertionIndex := sort.Search(len(q.waitingForAdd), func(i int) bool {
+		return waitEntry.readyAt.Before(q.waitingForAdd[i].readyAt)
+	})
+
+	tail := q.waitingForAdd[insertionIndex:]
+	q.waitingForAdd = append(q.waitingForAdd[:insertionIndex], waitEntry)
+	q.waitingForAdd = append(q.waitingForAdd, tail...)
+
+	q.notifyAt(waitEntry.readyAt)
+}
+
+// maxWait keeps a max bound on the wait time.  It's just insurance against weird things happening.
+// Checking the queue every 10 seconds isn't expensive and we know that we'll never end up with an
+// expired item sitting for more than 10 seconds.
+const maxWait = 10 * time.Second
+
+// waitingLoop runs until the workqueue is shutdown and keeps a check on the list of items to be added.
+func (q *delayingType) waitingLoop() {
+	for {
+		if q.shuttingDown {
+			return
+		}
+
+		func() {
+			q.waitingCond.L.Lock()
+			defer q.waitingCond.L.Unlock()
+			q.waitingCond.Wait()
+
+			if q.shuttingDown {
+				return
+			}
+
+			q.waitingLock.Lock()
+			defer q.waitingLock.Unlock()
+
+			// always check at least every 10 seconds
+			nextReadyCheck := q.clock.Now().Add(maxWait)
+			nextItemIndex := 0
+
+			for i, queuedItem := range q.waitingForAdd {
+				nextItemIndex = i
+				if queuedItem.readyAt.After(nextReadyCheck) {
+					nextReadyCheck = queuedItem.readyAt
+					break
+				}
+
+				q.Type.Add(queuedItem.data)
+			}
+
+			q.waitingForAdd = q.waitingForAdd[nextItemIndex:]
+			q.notifyAt(nextReadyCheck)
+		}()
+	}
+}
+
+// notifyAt: if the requested nextReadyCheck is sooner than the current check, then a new go func is
+// spawned to notify the condition that the waitingLoop is waiting for after the time is up.
+func (q *delayingType) notifyAt(nextReadyCheck time.Time) {
+	q.nextCheckLock.Lock()
+	defer q.nextCheckLock.Unlock()
+
+	if nextReadyCheck.After(q.nextCheckTime) {
+		return
+	}
+
+	now := q.clock.Now()
+	duration := nextReadyCheck.Sub(now)
+	// Insurance against weird things happening.  The longest time we'll ever wait is a few seconds
+	// so any bug in calculation will get a second shot a few seconds later instead of stalling.
+	if duration < 0 || duration > maxWait {
+		duration = maxWait
+	}
+	q.nextCheckTime = now.Add(duration)
+	ch := q.clock.After(duration)
+
+	go func() {
+		<-ch
+
+		q.waitingCond.Broadcast()
```

Comment:
If I understand correctly:
1. every call to `q.waitingCond.Broadcast()` will unblock the waiting loop
2. that loop will add any expired items, then re-call `notifyAt` with `max(now+maxWait, item[0].readyAt)`
3. normally that would accumulate notifying goroutines, but `notifyAt` no-ops if the requested notify time is after `q.nextCheckTime`

It seems like that short-circuit would prevent us from registering a `notifyAt` after the first registered notification fires (since `q.nextCheckTime` would then be in the past, and all notifyAt requests would likely be after it)